### ISSUE TITLE
Seafile doesn't sync all objects in git repos

### ISSUE THREAD

**Issue by allo-** [haiwen/seafile#2677]
Many times when I put a git repo into a Seafile library, Seafile does not sync all objects. When I backup the repo and compare it afterward, some files in the object store are completely missing. I do not have a seafile-ignore.txt file in the library and no other idea why the files are ignored. Syncing other folders works without missing files.

**Comment by feiniks** [comment]
Hello @allo- , seafile will ignore uploading temporary files generated on linux by default, ending with ~. If you have such files in your local directory, these files will not be uploaded.

**Comment by allo-** [comment]
My problem are files in `.git/objects/XX/` which have hexadecimal coded hashes as names.

**Comment by feiniks** [comment]
Hello @allo- , how did you upload this directory?  When uploading through a browser, the system may ignore these hidden files of git. You can try to upload through the client.

**Comment by allo-** [comment]
No, I am working with the repo in a synced folder. On the other computer, git always crashes when I added commits.

When I create a tar of the `.git` folder I can extract it afterward on the other computer and find that it contains some files in the objects folder, which are not in the synced `.git/objects` folder. Not all are missing, but some.

With git clone I would have thought that hardlinks could be a problem, but the repo was not cloned on the same filesystem and after a new commit, the new objects are unique, so they are just regular files not linked to another file. I wonder if git creates them in another unusual way that Seafile doesn't always handle correctly. Maybe something like the files being created (too) fast after each other or created as temp file and then renamed?

**Comment by feiniks** [comment]
Hello @allo-, if the client does not upload the .git directory at all, it means that the client is working normally. This problem may be related to the working mechanism of git. .git is the metadata directory of git. Seafile client will only upload the current directory status to the server. Other clients may have synchronized an intermediate state. You'd better synchronize the git repository through the synchronization mechanism of git itself, and the .git directory does not need to be copied.

**Comment by allo-** [comment]
It does upload most of it and I want it to upload it (yeah, there are better ways to push git data around ... still I want it for some versioned folders). I am not talking about a shallow clone, but a full repo inside the .git folder of the project, so not only metadata, but the full repository content.

I am not aware of a blacklist that prevents syncing the git folder. If there is any it is not working very well and I would like to be able to turn it off.

**Comment by feiniks** [comment]
Hello @allo- ,  in my case, all the files under the .git directory can be uploaded normally. Can you provide a more complete test procedure? Or you can check seafile.log and events.log to see if there is an error during the upload process that prevents the corresponding file from being uploaded.

**Comment by allo-** [comment]
I can check the logs next time again. I've never seen an error (like having the sync error indicator on the tray icon). I'll look into the logs next time I sync from the other location. Are there any specific tests I can do then?

**Comment by allo-** [comment]
I've made a diff what's missing for a repo (difference between "find >files.txt" on both sides) and four objects are not synced:

```
-./.git/objects/85/f3f1d6066b1d7aacd0913c65a23fbef32bb250
-./.git/objects/ee/35b6ba1bcc42ad8ea05c1758e1bdb6312c6894
-./.git/objects/f6/d929a2e6b350ba72d01760120f2c7c4f078b77
-./.git/objects/f9/5443e70fb34583fc152871082c354ec745a3cf
```

The files are not different but completely missing on the side where the repository is broken. I think the files are not different from other objects in the repo. Maybe git writes the files in some way or sets timestamps that may be a bit in the past or similar problems that prevent the seafile client from detecting the change?

**Comment by feiniks** [comment]
Hello @allo-,  which version of the client are you using?

**Comment by allo-** [comment]
On the end where the new files exist (and weren't uploaded): 8.0.5 (Ubuntu)
On the receiving end: 9.0.4 (Debian Testing)
When doing a resync I see in the web interface that the few missing files are uploaded, so I think it is not a problem of the client downloading the files.

But I have the problems since a longer time with updates in the meantime and I think its only now the 8 -> 9 thing and may have happened in the other direction as well.

Forcing a resync of the library successfully adds the missing files to the repo. The problem is, that I am on the wrong computer to force the resync when I notice that a git repo doesn't work because of missing files.

**Comment by feiniks** [comment]
Hello @allo- , I think 8.0.5 is an older version, and there may be some bugs in this version that are causing problems. 

Please upgrade to the latest version and see if you still have problems.

**Comment by allo-** [comment]
It's from the Ubuntu Repos. I can upgrade next week to another version.

**Comment by allo-** [comment]
At the moment I don't have the issue. As I cannot reproduce currently I close here and may reopen or create a new issue if I have problems again.

**Comment by Alukardd** [comment]
Hello.

The problem is still persists.  I have two clients: `9.0.12` on Debian Sid and `9.0.13 (latest AppImage from seafile.com)` on Arch Linux.
As workaround I can confirm that manual Right-Click on the Library and `Sync now` action is works.

**Comment by allo-** [comment]
Can still confirm on 9.0.4 in Ubuntu noble. All content gets synced correctly, but `.git/objects/` makes problems. I am not sure if all objects are missing or only some, but currently I have in each commit missing objects.

Did someone test if it is filesystem dependent? I am using btrfs. Maybe it does something different with timestamps/inodes or some notification API the client uses to detect changes?

EDIT:
I am not sure about the versions, because I am syncing between 9.0.15 (Debian) and 9.0.4 (Ubuntu). In the last test the *Ubuntu* side didn't check in all files, I need to watch out if the Debian version uploads everything.

Were there notable changes between 9.0.4 and 9.0.15 that may be related? I can try to get the latest version onto the Ubuntu PC.

**Comment by allo-** [comment]
I checked the logs. Ubuntu uploaded 3 objects and all the metadata (like `COMMIT_EDITMSG`, `index`, etc.). After using `rsync` to fetch the rest, Debian uploaded 5 more objects (created by rsync).

Can it be a race condition? Maybe git creates too many files in a row? That could also explain why "Sync now" works (not tested by me, but I'd try to check next time I commit from the Ubuntu PC).

**Comment by feiniks** [comment]
Hello @allo- , 

Since seafile client on Linux uses inotify to detect file changes, it is possible that there are some compatibility issues between btrfs and inotify. Can you try reproducing the issue under ext4 to see if it still occurs?

**Comment by Alukardd** [comment]
@feiniks 
> Since seafile client on Linux uses inotify to detect file changes, it is possible that there are some compatibility issues between btrfs and inotify. Can you try reproducing the issue under ext4 to see if it still occurs?

The problem is persist with `Seafile-x86_64-9.0.15.AppImage` on `ext4` filesystem.

**Comment by feiniks** [comment]
Hello @Alukardd ，

When this issue occurs, are there any sync errors reported on the client side? Errors encountered during indexing might also cause this problem. You can right-click the tray icon and select “show file sync errors”.

**Comment by allo-** [comment]
In principle should btrfs be one of the most advanced filesystems featurewise with good inotify support. I don't think that is a general problem with the FS, but wanted to mention it since some libraries may be less tested on new filesystems. Anyway, @Alukardd is using ext4, so it's probably no filesystem issue.

Another thing beside creating multiple object files in a row (or possible in parallel?) I could imagine is that git may create the files as temporary files and then instantly move them to the correct folder as soon as they are written. I don't really know the git interna, just thinking about what may cause race conditions when finding which files to upload.

I don't see sync errors here. Manual sync doesn't work either, did you do a complete resync for the files to be synced?

**Comment by Alukardd** [comment]
> When this issue occurs, are there any sync errors reported on the client side? 

No errors were reported. And library in the client marked as `Synced`.

> Manual sync doesn't work either

Manual sync works for me (right-click on the potentially problem library and choose `Sync now`). This action usually add-sync some git files like `.git/objects/53/19b2f7dae686d5a...`

**Comment by feiniks** [comment]
Hello @Alukardd @allo- ,

Perhaps you can set the SEAFILE_DEBUG environment variable to Watch and then start the sync client. After setting this variable, the inotify events will be recorded in seafile.log. When the issue occurs, you can check whether there are event records for the corresponding files in the log. This will help us confirm whether the system did not generate the events or generated them but failed to process them correctly.
For example:
```
export SEAFILE_DEBUG=Watch
./Seafile-x86-64.AppImage
```

**Comment by Alukardd** [comment]
In seafile.log I see
```
[11/26/25 13:23:10] wt-monitor(350): Created .git/objects/55/02b2c2267f2e7ad0c5147b25b31ed2dd4962f3.
```
but no more lines for `Adding event: create/update` for this file..
And I got notification about actual synchronization of this file only after Manual syncing the Library. (without any additional info in the Log, I guess because Watch is not related to manual sync process)

**Comment by feiniks** [comment]
Hi,

Git creates metadata files using hard links in .git. In Linux, when a hard link is created using the link() system call or the ln command, inotify triggers two different types of events because this process both creates a new directory entry and modifies the file’s metadata.

Seafile client indexes files only after detecting the IN_CLOSE_WRITE event via inotify. However, files created via hard links generate only an IN_CREATE event and do not produce an IN_CLOSE_WRITE event, which causes such files not to be indexed. If the file were processed at the IN_CREATE stage, it could be indexed while it is still being written, leading to issues.

At present, we do not have plans to support this edge case.
